### PR TITLE
config "calculate % without abstains" (motions) (Issue #2182)

### PR DESCRIPTION
**Pull request by matakuka**
The 100% base in this option is calculated by (valid votes  - abstains), making it necessary to know the "abstains" beforehand (in case of assignments), or knowing which type of vote 'vote' is (in case of motions).

### CODE REVIEW

File: openslides/motions/static/js/motions/base.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ angular.module('OpenSlidesApp.motions', ['OpenSlidesApp.users'])
             },
             methods: {
                 // returns object with value and percent
-                getVote: function (vote) {
+                getVote: function (vote,type) {
```
Comment (emanuelschuetze) [inline openslides/motions/static/js/motions/base.js]: please add a blank after ",".

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 2]
```diff
@@ -145,7 +145,7 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
                   <tr>
                     <td class="icon">
                       <i class="fa fa-thumbs-up fa-2x"></i>
-                    <td ng-init="voteYes = poll.getVote(poll.yes)">
+                    <td ng-init="voteYes = poll.getVote(poll.yes,'yes')">
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: add a blank after ","

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 3]
```diff
@@ -169,7 +169,7 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
                   <tr>
                     <td class="icon">
                       <strong style="font-size: 26px">&empty;</strong>
-                    <td ng-init="voteAbstain = poll.getVote(poll.abstain)">
+                    <td ng-init="voteAbstain = poll.getVote(poll.abstain,'abstain')">
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: add a blank after ","

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 4]
```diff
@@ -190,7 +190,7 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
                   <tr>
                     <td class="icon">
                       <i class="fa fa-ban fa-lg"></i>
-                    <td ng-init="votesInvalid = poll.getVote(poll.votesinvalid)">
+                    <td ng-init="votesInvalid = poll.getVote(poll.votesinvalid,'votesinvalid')">
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: add a blank after ","

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 5]
```diff
@@ -181,7 +181,7 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
                   <tr>
                     <td class="icon">
                       <i class="fa fa-check fa-lg"></i>
-                    <td ng-init="votesValid = poll.getVote(poll.votesvalid)">
+                    <td ng-init="votesValid = poll.getVote(poll.votesvalid,'votesvalid')">
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: add a blank after ","

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 6]
```diff
@@ -202,7 +202,7 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
                   <tr class="total">
                     <td class="icon">
                       <strong style="font-size: 16px">&sum;</strong>
-                    <td ng-init="votesCast = poll.getVote(poll.votescast)">
+                    <td ng-init="votesCast = poll.getVote(poll.votescast,'votescast')">
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: add a blank after ","

**Comment by emanuelschuetze** [comment]
After we have merged #2189 I'll make a test with this pr.

**Comment by emanuelschuetze** [comment]
Please rebase this pull request on top of current master (I want to test it with the merged pr #2189)

File: openslides/motions/static/js/motions/base.js [hunk 7]
```diff
@@ -90,7 +90,18 @@ angular.module('OpenSlidesApp.motions', ['OpenSlidesApp.users'])
...
+                        console.log(this.abstain);
+			            percentNumber = Math.round(vote * 100 /(this.votesvalid - this.abstain) * 10) / 10;
+                        } else if (type == 'abstain'){
+                          percentNumber = "";
+                        } else if (type == 'valid'|| type=='invalid'){
+                          percentNumber = Math.round(vote * 100 /(this.votescast) * 10) / 10;
+                        }
                     }
+                    
+
```
Comment (emanuelschuetze) [inline openslides/motions/static/js/motions/base.js]: please remove these 2 empty lines.

File: openslides/poll/models.py [hunk 8]
```diff
@@ -70,6 +70,7 @@ def print_weight(self, raw=False):
 PERCENT_BASE_CHOICES = (
     {'value': 'WITHOUT_INVALID', 'display_name': ugettext_lazy('Only all valid votes')},
     {'value': 'WITH_INVALID', 'display_name': ugettext_lazy('All votes cast (including invalid votes)')},
+    {'value': 'WITHOUT_ABSTAIN', 'display_name':  ugettext_lazy('Only YES and NO values')},
```
Comment (emanuelschuetze) [inline openslides/poll/models.py]: use better: "Only Yes and No votes" for the new option.

Change the string "Only all valid votes" to "All valid votes (Yes/No/Abstain)". This should better explain the difference between the options.

And please change the order (descending):
* "Only Yes and No votes"
* "All valid votes (Yes/No/Abstain)"
* "All votes cast (including invalid votes)"
* "Disabled (no percents)"

**Comment by emanuelschuetze** [comment]
Retested:
* motion slide does not show percent values if new 'Yes/No'-config-option is selected.
* motion pdf creation fails if new 'Yes/No'-config-option is selected (error: "MotionPoll" object has no attribute 'abstain')

**Comment by matakuka** [comment]
The pdf printouts (assigment/ motions) still need work. Motion shows the total percent including abstains, assignment shows nothing.

**Comment by matakuka** [comment]
I looked wrong. Assignment pdf printout works as expected. Motion pdf printout is still wrong, but I'm lost in where the pdf gets its percents from.
Here is where I lose track:
[motions/pdf.py:111](https://github.com/OpenSlides/OpenSlides/blob/28584c62f6b4b582e6168df24371f16bbbd54a43/openslides/motions/pdf.py#L111) --> [poll/models.py:173-177](https://github.com/OpenSlides/OpenSlides/blob/28584c62f6b4b582e6168df24371f16bbbd54a43/openslides/poll/models.py#L173-L177) --> ?

File: openslides/poll/models.py [hunk 9]
```diff
@@ -66,10 +66,13 @@ def print_weight(self, raw=False):
...
         return print_value(self.weight, percent_base)
 
-
 PERCENT_BASE_CHOICES = (
-    {'value': 'WITHOUT_INVALID', 'display_name': ugettext_lazy('Only all valid votes')},
-    {'value': 'WITH_INVALID', 'display_name': ugettext_lazy('All votes cast (including invalid votes)')},
+    {'value': 'WITHOUT_ABSTAIN', 'display_name':  ugettext_lazy(
+        'Only Yes and No votes')},
+    {'value': 'WITHOUT_INVALID', 'display_name': ugettext_lazy(
+        'Only all valid votes (Yes/No/Abstain)')},
```
Comment (emanuelschuetze) [inline openslides/poll/models.py]: As I suggested in my last comment: Please remove "Only" from this string: "All valid votes (Yes/No/Abstain)" 
Should be consistent with the "All votes cast (..)" option

**Comment by emanuelschuetze** [comment]
Retested:
* Motion slide works now.
* Motion pdf is created now. For the right percent values (if yes/no config option is selected) you should look at the [print_value function in poll/models.py](https://github.com/OpenSlides/OpenSlides/blob/28584c62f6b4b582e6168df24371f16bbbd54a43/openslides/poll/models.py#L256). There, I think, you have to modifiy the percent output. Try a bit with debug outputs.

**Comment by matakuka** [comment]
I disabled the wrong percent values on the printouts and just print the raw numbers. 
@normanjaeckel: is there an easy way to access the poll's "abstains" at [poll/models.py](https://github.com/OpenSlides/OpenSlides/pull/2192/files#diff-63a1c59fe8a052b345cbb1d36a2d7ccaR134)? Votesvalid etc. are there, but I couldn't get "yes/no/abstain" (I want to subtract the abstains from the valid votes)

File: openslides/poll/models.py [hunk 10]
```diff
@@ -119,10 +122,19 @@ def print_votescast(self):
...
             base = 100 / float(self.votesvalid)
-        elif self.get_percent_base_choice() == "WITH_INVALID" and self.votescast and self.votescast > 0:
+        elif self.get_percent_base_choice() == "WITH_INVALID" and\
+                self.votescast and self.votescast > 0:
             base = 100 / float(self.votescast)
+        # elif self.get_percent_base_choice() == "WITHOUT_ABSTAIN" and\
+        #     self.votesvalid and self.votesvalid > 0:
+        #     try:
+        #         base = 100 / float(self.votesvalid - abstain)
+        #         #TODO get abstains from server
```
Comment (emanuelschuetze) [inline openslides/poll/models.py]: Ok, now, the motion/assignment pdf has no percent values (only absolute votes). Not a really nice workaround but ok for (as long as we move the pdfs to client side - then we can use the JS code for percent values).

@normanjaeckel What do you think? Maybe do you have a quick solution to get the abstain value on  server side?

Comment (normanjaeckel) [inline openslides/poll/models.py]: The `CollectDefaultVotesMixin` is a mixin that should work for different poll/vote class systems in the respective apps. In this case you can not access the abstain votes in this abstract layer because there is not guarantee that the concrete poll class (MotionPoll e. g.) has abstention votes at all. What you have to do is to override the method `get_percent_base` in both the `MotionPoll` class and the `AssignmentPoll` class. Then you can substract the abstention easily in motion and you have top decide what should happen in the different assignment cases.

**Comment by normanjaeckel** [comment]
@matakuka Please check whether overriding `get_percent_base()` in `MotionPoll` may help.

**Comment by normanjaeckel** [comment]
@matakuka What is the status of this pull request?

**Comment by matakuka** [comment]
Still on my list, but with lower priority. I'll check it in the next days/this weekend.

File: openslides/assignments/views.py [hunk 11]
```diff
@@ -497,32 +497,31 @@ def append_to_pdf(self, story):
                     cell.append(Paragraph(
                         "(%s)" % candidate.structure_level,
                         stylesheet['Ballot_option_suffix_YNA']))
+                if self.poll.yesnoabstain:
```
Comment (matakuka) [inline openslides/assignments/views.py]: the change of indentation on the following lines fixes an issue that haven't been seen before: if a candidate had a structure_level, they were printed without ballot options on the paper.

**Comment by matakuka** [comment]
It should work now as desired, the pdf printouts include the percentages as well.
However, as I dabbled in several files and on several data model levels and the "proper thinking about this issue" was two months ago, I'm not entirely sure if I didn't accidentally inserted bad or unnecessary code (especially maybe duplicating "abstain/votesabstain" variables) . Also, the combination of a new mode of votes and this new config option leave quite a lot of test cases, and I may not have covered them all.
@normanjaeckel @emanuelschuetze

**Comment by normanjaeckel** [comment]
Hm. I think we need a different approach. I can have a look at it, but not at the moment.

File: openslides/assignments/views.py (hunk 11 above)
Comment (matakuka) [inline openslides/assignments/views.py]: - I moved this to a different PR #2254

**Comment by normanjaeckel** [comment]
First we should discuss the goal of the issue. See #2182. I can specify my critics on the current proposal if necessary but I think we should focus on the different voting methods first.

**Comment by emanuelschuetze** [comment]
As discussed: for motion only...

**Comment by matakuka** [comment]
Done.

I reworked it for motions only.

**Comment by matakuka** [comment]
@emanuelschuetze for testing.
Note: pdf creation has incorrect percentage calculations in master/motions/site.js, this PR fixes it.

**Comment by matakuka** [comment]
ping @emanuelschuetze

File: openslides/motions/config_variables.py [hunk 12]
```diff
@@ -1,4 +1,6 @@
 from django.core.validators import MinValueValidator
+from django.utils.translation import ugettext_lazy
+
```
Comment (emanuelschuetze) [inline openslides/motions/config_variables.py]: please remove this line

File: openslides/motions/static/js/motions/site.js [hunk 13]
```diff
@@ -1113,7 +1114,7 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions', 'OpenSlid
             }
         }
 
-        // save motion
+        // save motion.
```
Comment (emanuelschuetze) [inline openslides/motions/static/js/motions/site.js]: please remove this "."

File: openslides/motions/static/templates/motions/slide_motion.html [hunk 14]
```diff
@@ -23,36 +23,37 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
...
-              <td ng-init="voteNo = poll.getVote(poll.no)">
+              <td ng-init="voteNo = poll.getVote(poll.no, 'no')">
                 <span class="result_label"><translate>No</translate>:</span>
                 <span class="result_value" >
                   {{ voteNo.value }} {{ voteNo.percentStr }}
                 </span>
-                <div ng-if="voteNo.percentNumber">
-                  <uib-progressbar value="voteNo.percentNumber" type="danger"></uib-progressbar>
+                <div ng-if="voteNo.percentNumber >= 0">
+                  <uib-progressbar value="voteNo.percentNumber"  type="danger"></uib-progressbar>
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/slide_motion.html]: remove the new blank

File: openslides/motions/static/templates/motions/slide_motion.html [hunk 15]
```diff
@@ -23,36 +23,37 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
...
                 <strong style="font-size: 26px">&empty;</strong>
-              <td ng-init="voteAbstain = poll.getVote(poll.abstain)">
+              <td ng-init="voteAbstain = poll.getVote(poll.abstain, 'abstain')">
                 <span class="result_label"><translate>Abstain</translate>:</span>
                 <span class="result_value">
                   {{ voteAbstain.value }} {{ voteAbstain.percentStr }}
                 </span>
-                <div ng-if="voteAbstain.percentNumber">
+                <div ng-if="config('motions_poll_100_percent_base') != 'WITHOUT_ABSTAIN' && 
+                         voteAbstain.percentNumber >= 0">
```
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/slide_motion.html]: is this config check really required? voteAbstain.percentNumber should be null/undefined if WITHOUT_ABSTAIN is selected.

File: openslides/poll/models.py [hunk 16]
```diff
@@ -66,10 +66,11 @@ def print_weight(self, raw=False):
...
 
-
 PERCENT_BASE_CHOICES = (
-    {'value': 'WITHOUT_INVALID', 'display_name': ugettext_lazy('Only all valid votes')},
-    {'value': 'WITH_INVALID', 'display_name': ugettext_lazy('All votes cast (including invalid votes)')},
+    {'value': 'WITHOUT_INVALID', 'display_name': ugettext_lazy(
+        'All valid votes (Yes/No/Abstain)')},
+    {'value': 'WITH_INVALID', 'display_name': ugettext_lazy(
+        'All votes cast (including invalid votes)')},
     {'value': 'DISABLED', 'display_name': ugettext_lazy('Disabled (no percents)')})
```
Comment (emanuelschuetze) [inline openslides/poll/models.py]: Please reformat it in one line. The lines are not yet to long.

**Comment by matakuka** [comment]
as requested, some minor changes & a rebase @emanuelschuetze

File: openslides/motions/config_variables.py [hunk 17]
```diff
@@ -23,6 +24,10 @@ def get_config_variables():
     papers' and 'PDF'. The generator has to be evaluated during app loading
     (see apps.py).
     """
+    percent_base_choices_motion = PERCENT_BASE_CHOICES
```
Comment (emanuelschuetze) [inline openslides/motions/config_variables.py]: please use capital letters for constants

**Comment by normanjaeckel** [comment]
Discussed some changes with @emanuelschuetze. Rest is ok.

File: openslides/motions/config_variables.py (hunk 17 above)
Comment (emanuelschuetze) [inline openslides/motions/config_variables.py]: done in #2331

**Comment by emanuelschuetze** [comment]
I created a new PR #2331 which makes last fixes for this feature. So I closed this PR here.
